pycbc/live/snr_optimizer.py

```diff
@@ -0,0 +1,383 @@
+    population = numpy.random.uniform(
+        low=0.0, high=1.0, size=(int(cli_args.pso_particles), 4)
+    )
+    population = normalize_population(population, min_bounds, max_bounds)
+
+    if cli_args.include_candidate_in_optimizer:
+        # add the initial point to the population
+        population = numpy.concatenate((population[:-1],
+                                        initial_point))
+
+    optimizer = ps.single.GlobalBestPSO(
+        n_particles=int(cli_args.pso_particles),
+        dimensions=4,
+        options=options,
+        bounds=(min_bounds, max_bounds),
+        init_pos=population
+    )
+    _, results = optimizer.optimize(
+        compute_minus_network_snr_pso,
+        iters=int(cli_args.pso_iters),
+        n_processes=cli_args.cores,
+        args=extra_args
+    )
+    return results
+
+
+optimize_funcs = {
+    'differential_evolution': optimize_di,
+    'shgo': optimize_shgo,
+    'pso': optimize_pso
+}
+
+# The following sets the default values of the options, but allows us to check
+# if the option has been given on the command line
+
+# For each optimizer, we have a dictionary of the options, its help
+# message and default value
+
+option_dict = {
+    'differential_evolution': {
+        'maxiter': ('The maximum number of generations over which the entire '
+             'population is evolved.', 50),
+        'popsize': ('A multiplier for setting the total population size.',
+                    100),
+    },
+    'shgo': {
+        'samples': ('Number of sampling points used in the construction of '
+                    'the simplicial complex.', 76),
+        'iters': ('Number of iterations used in the construction of the '
+                  'simplicial complex.', 3),
+    },
+    'pso': {
+        'iters': ('Number of iterations used in the particle swarm '
+                  'optimization.', 5),
+        'particles': ('Number of particles used in the swarm.', 250),
+        'c1': ('The hyperparameter c1: the cognitive parameter.', 0.5),
+        'c2': ('The hyperparameter c2: the social parameter.', 2.0),
+        'w': ('The hyperparameter w: the inertia parameter.', 0.01),
+    }
+}
+
+def insert_snr_optimizer_options(parser):
+    opt_opt_group = parser.add_argument_group("SNR optimizer configuration options.")
+    # Option to choose which optimizer to use:
+    optimizer_choices = sorted(list(option_dict.keys()))
+    opt_opt_group.add_argument('--optimizer',
+        type=str,
+        default='differential_evolution',
+        choices=optimizer_choices,
+        help='The optimizer to use, choose from ' + ', '.join(optimizer_choices))
+
+    # Add the generic options
+    opt_opt_group.add_argument('--include-candidate-in-optimizer',
+        action='store_true',
+        help='Include parameters of the candidate event in the initialized array for the optimizer. '
+             'Only relevant for --optimizer pso or differential_evolution')
+    opt_opt_group.add_argument('--seed',
```

Comment:
As this is also going to go through to the pycbc_live options, it would be good to update this to be `--optimizer-seed` (or similar), just to be obvious that it is going to be passed through

Reply:
Needed to actually add the seed in too, do you know if setting the seed in `pycbc_optimize_snr` will pass that numpy arg seed through to `snr_optimizer`?

Reply:
I dont think so, the seed was added after I did the initial options change

Reply:
(though you can test this by running twice and checking outputs are the same)

Reply:
They have the same outputs when running twice so I'm going to say it is passed through.